include/osquery/expected.h

```diff
@@ -0,0 +1,164 @@
+//
+//  expected.hpp
+//  gmock
+//
+//  Created by Max Kareta on 5/15/18.
+//
+
+#pragma once
+
+#include <include/osquery/error.h>
+
+/** Utility class that should be used in function that return
+ * either error or value. Expected enforce developer to test for success and
+ * check error if any.
+ *
+ * ExpectedValue:
+ * ExpectedValue<std::string> function() {
+ *   if (test) {
+ *    return "ok";
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * Expected:
+ * Expected<PlatformProcess> function() {
+ *   if (test) {
+ *    return std::unique_prt(new PlatformProcess(pid));
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * auto result = function();
+ * if (result) {
+ *   ...use *result
+ * } else {
+ *   auto error = result->getError();
+ * }
+ */
+
+namespace osquery {
+
+class ExpectedBase {
+ public:
+  ExpectedBase(Error* error) = delete;
+  ExpectedBase(Error error) = delete;
+  ExpectedBase() : error_(nullptr), hasError_(false) {}
+  ExpectedBase(std::shared_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+  ExpectedBase(std::unique_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+
+  ~ExpectedBase() {
+    assert(errorChecked_ && (!hasError_ || errorUsed_));
+  }
+
+  std::shared_ptr<Error> getError() {
+    errorUsed_ = true;
+    errorChecked_ = true;
+    return error_;
+  }
+
+  explicit operator bool() {
+    errorChecked_ = true;
+    return !hasError_;
+  }
+
+ protected:
+  std::shared_ptr<Error> error_;
+  bool hasError_;
+  bool errorChecked_;
+  bool errorUsed_;
+};
+
+template <class T>
+class ExpectedValue : public ExpectedBase {
+ private:
+  static const bool isPointer = std::is_pointer<T>::value;
+  static_assert(!isPointer, "Use Expected class for pointers");
+  using reference = typename std::remove_reference<T>::type&;
+  using const_reference = const typename std::remove_reference<T>::type&;
+  using pointer = typename std::remove_reference<T>::type*;
+  using const_pointer = const typename std::remove_reference<T>::type*;
```

Comment:
To avoid code duplication and to make it more explicit you can use `std::add_const`, `std::add_pointer`, `std::add_lvalue_reference` from [type_traits](https://en.cppreference.com/w/cpp/header/type_traits):
```
  using value = typename std::remove_reference<T>::type;
  using reference = typename std::add_lvalue_reference<value>::type;
  using const_reference = typename std::add_const<reference>::type;
  using pointer = typename std::add_pointer<value>::type;
  using const_pointer = typename std::add_const<pointer>::type;
```

And to avoid of possible build problems in case of some code refactoring it is better to include `<type_traits>` to this file.